tests/qml/helper.cpp

```diff
@@ -0,0 +1,111 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "helper.h"
+
+#include "l18nstrings.h"
+#include "mozillavpn.h"
+#include "settingsholder.h"
+
+TestHelper::TestHelper() {
+  FeatureList::instance()->initialize();
+  m_whatsNewModel = new WhatsNewModel();
+}
+
+TestHelper* TestHelper::instance() {
+  static TestHelper* s_instance = new TestHelper();
+  return s_instance;
+}
+
+bool TestHelper::mainWindowLoadedCalled() const {
+  return m_mainWindowLoadedCalled;
+}
+
+void TestHelper::setMainWindowLoadedCalled(bool val) {
+  m_mainWindowLoadedCalled = val;
+}
+
+void TestHelper::triggerInitializeGlean() const {
+  emit MozillaVPN::instance()->initializeGlean();
+}
+
+void TestHelper::triggerRecordGleanEvent(const QString& event) const {
+  emit MozillaVPN::instance()->recordGleanEvent(event);
+}
+
+void TestHelper::triggerSendGleanPings() const {
+  emit MozillaVPN::instance()->sendGleanPings();
+}
+
+void TestHelper::triggerSetGleanSourceTags(const QStringList& tags) const {
+  emit MozillaVPN::instance()->setGleanSourceTags(tags);
+}
+
+void TestHelper::triggerAboutToQuit() const {
+  emit MozillaVPN::instance()->aboutToQuit();
+}
+
+// The next two functions duplicate mozillavpn.h logic, but I'm not
+// sure how else to generate an "expected" value for tst_mainWindowGlean.qml
+// as it needs to be dynamic based on the platform people are running tests on.
+// At the least this should fail give someone pause for thought if they change
+// the mozillavpn.h logic.
+QString TestHelper::osVersion() const { return QSysInfo::productVersion(); }
```

Comment:
@bakulf github might swallow our conversation because i fixed a different problem and I'd like to understand.

You wrote:
> Can you use MozillaVPN as a singleton and write this as: `QString TestHelper::osVersion() const { return MozillaVPN::instance()->osVersion(); }` ?

I don't think so because the MozillaVPN::instance() is the mocmozillavpn.cpp implementation, no?

If you know of a way of me being able to use the real mozillavpn instance that would be awesome.

Reply:
you don't need to use the "real" mozillavpn instance. There is only 1 instance, which is yours in this QML app.
But if you see, your mocmozillavpn.cpp implements the class definition from 'mozillavpn.h', and the `osVersion` implementation comes from this header file.

If you are unable to replace this with:
`QString TestHelper::osVersion() const { return MozillaVPN::instance()->osVersion(); }`

we have a bug somewhere else.